include/circt/Dialect/HW/CustomDirectiveImpl.h

```diff
@@ -0,0 +1,83 @@
+//===- CustomDirectiveImpl.h - Table-gen custom directive impl --*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides common custom directives for table-gen assembly formats.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_HW_CUSTOMDIRECTIVEIMPL_H
+#define CIRCT_DIALECT_HW_CUSTOMDIRECTIVEIMPL_H
+
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/OpImplementation.h"
+
+namespace circt {
+
+//===----------------------------------------------------------------------===//
+// InputPortList Custom Directive
+//===----------------------------------------------------------------------===//
+
+/// Parse a list of instance input ports.
+/// input-list ::= `(` ( input-element (`,` input-element )* )? `)`
+/// input-element ::= identifier `:` value `:` type
+ParseResult
+parseInputPortList(OpAsmParser &parser,
+                   SmallVector<OpAsmParser::UnresolvedOperand, 4> &inputs,
+                   SmallVector<Type, 1> &inputTypes, ArrayAttr &inputNames);
+
+/// Print a list of instance input ports.
+void printInputPortList(OpAsmPrinter &p, Operation *op, OperandRange inputs,
+                        TypeRange inputTypes, ArrayAttr inputNames);
+
+//===----------------------------------------------------------------------===//
+// OutputPortList Custom Directive
+//===----------------------------------------------------------------------===//
+
+/// Parse a list of instance output ports.
+/// output-list ::= `(` ( output-element (`,` output-element )* )? `)`
+/// output-element ::= identifier `:` type
+ParseResult parseOutputPortList(OpAsmParser &parser,
+                                SmallVector<Type, 1> &resultTypes,
+                                ArrayAttr &resultNames);
```

Comment:
If this can truly only have one output type, then should this be `Type &resultType`?

Reply:
There as many types as the input port list has values. This was just required because the auto-generated parser declares a `SmallVector<Type, 1>`. `SmallVectorImpl` solves this.